designs/authentication-use-cases.md

```diff
@@ -0,0 +1,96 @@
+# Authentication Use Cases
+
+As a technical specification, Webauthn has "many ways" it case be used that *appear* to all be
+valid. However, there is an unwritten set of combinations that are intended, and combinations
+that can both cause confusion to users or allows verification bypasses at worst.
+
+This document details the use cases that we support to understand how we should structure our library
+to ensure it can never be used/held incorrectly.
+
+> NOTE: This is not intended to be a simple introduction to webauthn, and will assume extensive prior
+> knowledge.
+
+## Key Terms
+
+* Non-discoverable credential - https://www.w3.org/TR/webauthn-2/#credential-id
+
+This is the "common" type of credential that is used. These are commonly implemented with a scheme
+known as "key-wrapped-key". This is where the authenticator has a single private key, and encrypts
+a credential with that key. The credential ID is the encrypted credential.
+
+For the authenticator to then operate, it must be presented with the credential id, that the authenticator
+decrypts and then uses in the authentication ceremony.
+
+* Discoverable Credentials - https://www.w3.org/TR/webauthn-2/#client-side-discoverable-credential
+
+A discoverable credential is previously called a resident key. This is when the *client* holds the
+ability to find credentials locally rather than requiring them to be supplied in the allowedCredentials
+set in authentication.
+
+* User Verification - https://www.w3.org/TR/webauthn-2/#user-verification
+
+The process of supplying extra or supplementary authentication to the authenticator. This improves
+the authentication from "there is a person present" to "this specific owner of the authenticator" is
+present. Conseder UV=True as the authenticator is a self container MFA device, where UV=false means that
+the device is a SFA, and other authentication is required to compose a complete MFA.
+
+At registration we store the policy and verification that was supplied in the internal Credential
+type so that later decisions can be made for security policy.
+
+## Authentication Work Flows
+
+### Homogenous Non-Discoverable Credentials
+
+This is a set of credentials with a homogeneous UV policy - either all discouraged or all required.
+
+This policy is derived from the set of UV=true/false flags in the allow credential list. If the
+UV flags are inconsistent an error is raised.
+
+In the case all of these credentials are discouraged, we still assert that the UV bit matches
+the returned authentication as it *is* valid for UV=true even when discouraged and we store that
+in the credential.
+
+* Example
+
+A laptop where the user has multiple UV=false yubikeys that may be used as a authentication factor.
+
+A user indicates they want to authenticate with their touchid as a self contained MFA. The user owns
```

Comment:
in this case it could be a mix of credentials: yubikeys + touchid that the RP would want to accept right?

Reply:
Yep, absolutely. So yubikey + touchid, and the touchid could be uv=true at registration, but we use the uv policy at registration for consistency I think. So if we register with uv=discouraged, but we get back uv=true from the device, then we still assert that holds true. 

I think maybe i'll update this section with some more detail examples of the registration through to auth flow though. I wonder if this could make uv=preferred at registration, and then we just check each credentials uv=true/false would then be viable for these situations.